core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1413,55 +1413,106 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
       n: Int,
       timeout: FiniteDuration
   )(implicit F: Temporal[F2]): Stream[F2, Chunk[O]] =
-    this
-      .covary[F2]
-      .pull
-      .timed { timedPull =>
-        def resize(c: Chunk[O], s: Pull[F2, Chunk[O], Unit]): (Pull[F2, Chunk[O], Unit], Chunk[O]) =
-          if (c.size < n) s -> c
-          else {
-            val (unit, rest) = c.splitAt(n)
-            resize(rest, s >> Pull.output1(unit))
+    fs2.Stream.force {
+      for {
+        demand <- Semaphore[F2](n.toLong)
+        supply <- Semaphore[F2](0L)
+        buffer <- Ref[F2].of(
+          JunctionBuffer[O](Vector.empty[O], endOfSupply = None, endOfDemand = None)
+        )
+      } yield {
+        def enqueue(t: O): F2[Boolean] =
+          for {
+            _ <- demand.acquire
+            buf <- buffer.modify(buf => (buf.copy(buf.data :+ t), buf))
+            _ <- supply.release
+          } yield buf.endOfDemand.isEmpty
+
+        def waitN(s: Semaphore[F2]) =
+          F.guaranteeCase(s.acquireN(n.toLong)) {
+            case Succeeded(_) => s.releaseN(n.toLong)
+            case _            => F.unit
+          }
+
+        def acquireSupplyUpToNWithin(n: Long, timeout: FiniteDuration): F2[Long] =
```

Comment:
Since this `timeout` is the same as the parameter of the `main` function, we can remove the parameter here.
```suggestion
        def acquireSupplyUpToNWithin(n: Long): F2[Long] =
```

Reply:
Done